src/Compilers/Core/Portable/CodeAnalysisEventSource.cs

```diff
@@ -38,12 +37,63 @@ public static class Tasks
         internal void StartSingleGeneratorRunTime(string generatorName, string assemblyPath, string id) => WriteEvent(3, generatorName, assemblyPath, id);
 
         [Event(4, Message = "Generator {0} ran for {2} ticks", Keywords = Keywords.Performance, Level = EventLevel.Informational, Opcode = EventOpcode.Stop, Task = Tasks.SingleGeneratorRunTime)]
-        internal void StopSingleGeneratorRunTime(string generatorName, string assemblyPath, long elapsedTicks, string id) => WriteEvent(4, generatorName, assemblyPath, elapsedTicks, id);
+        internal void StopSingleGeneratorRunTime(string generatorName, string assemblyPath, long elapsedTicks, string id)
+        {
+            unsafe
+            {
+                fixed (char* generatorNameBytes = generatorName)
```

Comment:
💭 I think I prefer this form as a way to ensure buffer lengths always match up:

```cs
fixed (char* generatorNameBytes = generatorName)
fixed (char* assemblyPathBytes = assemblyPath)
fixed (char* idBytes = id)
{
    Span<EventData> data = stackalloc EventData[]
    {
        new() { DataPointer = (IntPtr)generatorNameBytes, Size = (generatorName.Length + 1) * sizeof(char) },
        new() { DataPointer = (IntPtr)assemblyPathBytes, Size = (assemblyPath.Length + 1) * sizeof(char) },
        new() { DataPointer = (IntPtr)(&elapsedTicks), Size = sizeof(long) },
        new() { DataPointer = (IntPtr)idBytes, Size = (id.Length + 1) * sizeof(char) },
    };

    fixed (EventData* dataPtr = data)
    {
        WriteEventCore(eventId: 4, data.Length, dataPtr);
    }
}
```

Reply:
Could go one step further:

```cs
fixed (char* generatorNameBytes = generatorName)
fixed (char* assemblyPathBytes = assemblyPath)
fixed (char* idBytes = id)
{
    Span<EventData> data = stackalloc EventData[]
    {
        eventDataForString(generatorName, generatorNameBytes),
        eventDataForString(assemblyPath, assemblyPathBytes),
        eventDataForInt64(&elapsedTicks),
        eventDataForString(id, idBytes),
    };

    fixed (EventSource.EventData* dataPtr = data)
    {
        WriteEventCore(eventId: 4, data.Length, dataPtr);
    }
}

static EventData eventDataForString(string value, char* ptr)
{
    fixed (char* ptr2 = value)
    {
        if (ptr2 != ptr)
            throw new ArgumentException("Pinned value must match string.");
    }

    return new EventData()
    {
        DataPointer = (IntPtr)ptr,
        Size = (value.Length + 1) * sizeof(char),
    };
}

static EventData eventDataForInt64(long* ptr)
{
    return new EventData()
    {
        DataPointer = (IntPtr)ptr,
        Size = sizeof(long),
    };
}
```

Reply:
Can you take a look at the 2nd commit. Followed your suggestions fairly closely as it's a lot better than the framework's EventSource code that I originally mimic'ed.